src/main/java/com/hazelcast/gcp/GcpClient.java

```diff
@@ -58,6 +59,23 @@
             return gcpConfig.getProjects();
         }
         LOGGER.finest("Property 'projects' not configured, fetching the current GCP project");
+
+        try {
+            String projectFromMetadataApi = gcpMetadataApi.currentProject();
```

Comment:
I think this code execution should not be here.

1. What is the case when VM does not have access to its Metadata? When user removes explicitly?
2. Is it possible to check the HTTP Code instead of parsing the message? I guess GCP Metadata should return something like: `403 (Forbidden)`.
3. The code responsible for this should be in `GcpMetadataApi` and it could throw some exception like `HttpForbiddenException` (you can create it, if there is nothing like this in the Java lib). Then this exception should be understood by the `RetryUtils` and not retried (you can add a parameter to `RetryUtils.retry()` )

So the execution would be something like this:

```
        return singletonList(RetryUtils.retry(new Callable<String>() {
            @Override
            public String call() {
                return gcpMetadataApi.currentProject();
            }
        }, RETRIES), Arrays.asList(HttpForbiddenException.class));
```

Reply:
@leszko  Thanks for the feedback! I have started to incorporate some of the updates. I guess the only way it wouldn't be able to query for its metadata is if it did not have API access scope for Compute Engine of at least read-only, but in that case it hits the permissions error on the access token calls before it gets to this method anyway, so I'll refactor this. Is there an easy way I can know if someone is trying to use the plugin from within the GCP network where the hazelcast cluster is running vs. outside of the GCP network? Because right now in the GCPClient.projectFromConfigOrMetadataApi method, it first checks the configuration, and if it is empty, will try to retrieve the project from the metadata. So, in the case where the user is trying to use the plugin outside of the GCP network and they don't specify a project name in the configuration, the gcpMetadataApi.currentProject() method will actually return a response that is 200 (OK) rather than a 403(Forbidden), which is why I initially was thinking of just parsing the message to determine whether I should log the error. Ideally, I would be able to log the error before it even tries to get the project from the metadata (in the case of a client outside the GCP network). This situation is mirrored with the zonesFromConfigOrMetadataApi method. Any thoughts on the best way to do this?

Reply:
@ryanlindeborg I think there is no way to check if you're running inside GCP (VM Machine) or outside. Nothing other than trying to reach Metadata API.

There is one thing I don't get, why do you say that `gcpMetadataApi.currentProject()` returns HTTP 200 if run from outside GCP? I'd expect it throws an exception.

One more thing, try to make everything simple, if this error-checking really complicate the things and make additional calls to GCP API, we may want to decide that it's not worth the trade-off. 

Thanks again for working on that!

Reply:
@leszko Ok will do! When I run from a client outside GCP, the gcpMetadataApi.currentProject() returns this with a status code of 200:
`<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><html><head><meta http-equiv="refresh" content="0;url=http://finder.cox.net/main?ParticipantID=96e687opkbv4scrood8k84drs6gw5duf&FailedURI=http%3A%2F%2Fmetadata.google.internal%2FcomputeMetadata%2Fv1%2Fproject%2Fproject-id&FailureMode=1&Implementation=&AddInType=4&Version=pywr1.0&ClientLocation=us"/><script type="text/javascript">url="http://finder.cox.net/main?ParticipantID=96e687opkbv4scrood8k84drs6gw5duf&FailedURI=http%3A%2F%2Fmetadata.google.internal%2FcomputeMetadata%2Fv1%2Fproject%2Fproject-id&FailureMode=1&Implementation=&AddInType=4&Version=pywr1.0&ClientLocation=us";if(top.location!=location){var w=window,d=document,e=d.documentElement,b=d.body,x=w.innerWidth||e.clientWidth||b.clientWidth,y=w.innerHeight||e.clientHeight||b.clientHeight;url+="&w="+x+"&h="+y;}window.location.replace(url);</script></head><body></body></html>`

Reply:
I see, I wasn't aware of it. So, yes, we need to parse it somewhere and give the error to the user. Good finding btw.!